src/Polly.Core/Retry/RetryResilienceStrategy.cs

```diff
@@ -53,6 +53,11 @@ protected internal override async ValueTask<Outcome<T>> ExecuteCore<TState>(Func
         {
             var startTimestamp = _timeProvider.GetTimestamp();
             var outcome = await StrategyHelper.ExecuteCallbackSafeAsync(callback, context, state).ConfigureAwait(context.ContinueOnCapturedContext);
```

Comment:
One thing to be aware. If the outcome with disposable result is produced and then you are replacing it with exception, it might lead to a memory leak. 

I think it might be good idea to encapsulate all this handling into `ExecuteCallbackSafeAsync` method.

Reply:
If we would move the code inside the `ExecuteCallbackSafeAsync` then we don't have access to the `isLastAttempt`. @kmcclellan suggested to respect the cancellation request only under certain circumstances. 

Please see the other comment section for more details.

Reply:
> If we would move the code inside the ExecuteCallbackSafeAsync then we don't have access to the isLastAttempt.

The `ExecuteCallbackSafeAsync` is internal, we could pass that information to the function. Wdyt?

Reply:
That function is used by almost all strategies, where the "retry" concept may or may not make any sense. Of cource we can pass true as a default `isLastAttempt` but it feels a bit awkward to me.

Reply:
Are you still suggesting to add a new `isLastAttempt` parameter with default value? I'm fine with either approach, I just want to close this PR this year if possible 😁

Reply:
Aaah sorry about this, got busy with other stuff. I would say meh, we can add it, it's internal detail anyway. At least this logic will be encapsulated there.

So a new default value with `isLastAttempt  = true`. Retry strategy will provide its own value.

Reply:
Hi! I am the reporter of the bug. Hoping I can help move this forward to the right solution!

The concern about disposable outcomes being thrown away is not really relevant to this change. Abiding by the rules of cancellation, we should _not_ replace an outcome unless we have more work to do (in which case it was probably an exception and won't have a result to dispose).

> Don’t throw OperationCanceledException after you’ve completed the work, just because the token was signaled

It is only _handled outcomes that are not the last outcome_ which could have a disposable result thrown away. This is true whether we attempt another execution or acknowledge cancellation. It's fair to say that retry policies currently don't support handling disposable outcomes.

I'm not even sure what `ExecuteCallbackSafeAsync` would supposedly do with `isLastAttempt`. In order to dispose, it would also need to know that the strategy would handle the outcome.